src/libsyntax/errors/emitter.rs

```diff
@@ -361,16 +361,28 @@ fn print_diagnostic(dst: &mut Destination,
                     msg: &str,
                     code: Option<&str>)
                     -> io::Result<()> {
-    if !topic.is_empty() {
-        dst.start_attr(term::Attr::ForegroundColor(lvl.color()))?;
-        write!(dst, "{}: ", topic)?;
-        dst.reset_attrs()?;
-    }
 
-    dst.start_attr(term::Attr::Bold)?;
-    write!(dst, "{}: ", lvl.to_string())?;
-    write!(dst, "{}", msg)?;
-    dst.reset_attrs()?;
+
+    match lvl {
+        Level::Error => {
```

Comment:
That's true, special casing warning is probably ok. There are some cases where we have embedded warnings though. If we can do the header thing I think it'd be *better* (doesn't seem ... that hard).